pkg/cli/image/mirror/mirror.go

```diff
@@ -168,7 +168,7 @@ func NewCmdMirrorImage(streams genericclioptions.IOStreams) *cobra.Command {
 	flag.BoolVar(&o.SkipMount, "skip-mount", o.SkipMount, "Always push layers instead of cross-mounting them")
 	flag.BoolVar(&o.SkipMultipleScopes, "skip-multiple-scopes", o.SkipMultipleScopes, "Some registries do not support multiple scopes passed to the registry login.")
 	flag.BoolVar(&o.Force, "force", o.Force, "Attempt to write all layers and manifests even if they exist in the remote repository.")
-	flag.BoolVar(&o.KeepManifestList, "keep-manifest-list", o.KeepManifestList, "If an image is part of a manifest list, always mirror the list even if only one image is found. The default is to mirror the specific image unless unless --filter-by-os is '.*'.")
+	flag.BoolVar(&o.KeepManifestList, "keep-manifest-list", o.KeepManifestList, "If an image is part of a manifest list, always mirror the list even if only one image is found. The default is to mirror the specific image unless unless --filter-by-os is passed.")
```

Comment:
What happens if `--filter-by-os` is set to something that matches multiple targets within the manifest list?  I dunno if anyone has a spec around manifest-list severability, but it would be nice if you could have the list and then *n* of *m* referenced children, still be able to push the manifest list up into the target registry, and only get pull errors if someone tried to use the list to get at a referenced image that had not been mirrored.

Reply:
You're stretching the limits of what I know about manifestLists, only way i'll truly be able to answer this is if you have an example of such a manifestList that I can try :)

Reply:
`registry.redhat.io/openshift4/ose-must-gather:latest` is an example of a multi-arch image (see inspection [here][1]), if you want one to play around with.

[1]: https://github.com/openshift/oc/pull/627#issuecomment-718924035

Reply:
I meant, a manifestList with multiple targets of same os/arch. I've been using `quay.io/prometheus/prometheus:v2.18.1` multi-arch image in my testing.  I see the must-gather image does not have multiple targets for same os/arch? 
```
  OS            DIGEST
  linux/amd64   sha256:47c2f751ab0d5ee88e2826749f1372e6a24db3d0c0c942136ae84db17cb7f086
  linux/ppc64le sha256:70ebe4f2c01e17b84617c58dd5d64b00fe91bf6a3bc216b6d9f4690552f48265
  linux/s390x   sha256:aca00f9e47286c74b5b5d9ac7b3b130b9ce3ee087cb940269a0130aa7d569d18
```

Reply:
`filter-by-os` matches `os/arch` not just `os` - is that what you mean?

Reply:
So I mean if you ran `oc image mirror --filter-by-os 'amd64|ppc64le quay.io/prometheus/prometheus:v2.18.1 ...` or whatever, to match both `linux/amd64` and `linux/ppc64le`.  Ideally that would bring over the manifest list, the `linux/amd64` image, and the `linux/ppc64le` image, but not the `linux/390x` image.  It's not clear to me if registries and other tooling would support the original manifest list in that situation (because it would also reference the `linux/s390x` image we were not mirroring), or if you'd need to create a new manifest-list image (breaking signatures on the manifest-list image) that only referenced the images you were mirroring.

Reply:
This change is to always mirror the manifest list along with any filtered os/arch manifests. The only way to _not_ mirror the manifest list  is to run something like:
```console
$ oc image mirror registry.redhat.io/openshift4/ose-must-gather:latest=localhost:5000/mustgather
```
and you'll get the warning letting you know you did not mirror the manifest list, only the manifest for your system os/arch:
```console
W1130 22:57:41.622731   49734 manifest.go:414] Chose linux/amd64 manifest from the list sha256:d8e0b62c75bff609394bfaa43c923480d2d0bf2f7a3f36fd28ec601f99f3d25c
To include the manifest list digest: sha256:d8e0b62c75bff609394bfaa43c923480d2d0bf2f7a3f36fd28ec601f99f3d25c, use --keep-manifest-list=true
```

Reply:
and these will mirror the manifest list along w/ your filtered os/arch manifests:
```console
$ oc image mirror registry.redhat.io/openshift4/ose-must-gather:latest=localhost:5000/mustgather --keep-manifest-list=true
$ oc image mirror registry.redhat.io/openshift4/ose-must-gather:latest=localhost:5000/mustgather --filter-by-os=any/valid
```

Reply:
`--filter-by-os` will construct a manifest list with only the filtered manifests listed, therefore changing the digest of the mirrored manifestlist (or if only one matches it will only mirror the one and not mirror the manifestlist at all)

`--keep-manifest-list` was intended to preserve the entire manifestlist, regardless of what manifests have been filtered via `--filter-by-os`.

Without doing the forensics to confirm, I suspect that keep-manifest-list was implemented without the knowledge that that `--filter-by-os` could rewrite the manifestlist. This probably went unnoticed because most of the images mirrored with the `--keep-manifest-list` flag only contained a single entry, so the digest would remain the same.

The least surprising solution would be (IMO) to not touch `--filter-by-os` behavior at all unless used in conjunction with `--keep-manifest-list`, i.e.:

`oc image mirror --filter-by-os=<filter>`: same behavior as today
`oc image mirror --filter-by-os=<filter> --keep-manifest-list=true`: manifest list is not modified and mirrored, and any manifests matching the filter are mirrored.